Fiber test renderer

iamdustan opened the pull request:
This is the start of reimplementing `ReactTestRenderer` on Fiber. Current tests status:

- [x] renders a simple component (2ms)
- [x] renders a top-level empty component
- [x] exposes a type flag (2ms)
- [x] can render a composite component (1ms)
- [x] renders some basics with an update
- [x] exposes the instance (1ms)
- [x] updates types (1ms)
- [x] updates children
- [x] does the full lifecycle (1ms)
- [x] gives a ref to native components (2ms)
- [x] warns correctly for refs on SFCs
- [x] allows an optional createNodeMock function
- [x] supports unmounting when using refs
- [x] supports unmounting inner instances (3ms)
- [x] supports updates when using refs
- [x] supports error boundaries
- [x] compatibility: add support for `unstable_batchedUpdates`

A good chunk of those tests have to do with refs and the ability to provide a `createNodeMock` function. I believe that `ReactFiberReconciler` will need to be updated with a `getPublicInstance` method to handle this.

Ignore all changes in `ReactTestRenderer-test.js`. The first is to import the fiber renderer instead of stack (actually, is there a better way to do this? should I copy the ReactTestRenderer-test for ReactTestFiberRenderer?), the rest are testing/skip/etc and will be backed out when done.

cc @spicyj

aweary commented:
> should I copy the ReactTestRenderer-test for ReactTestFiberRenderer?

👍 I think we'll want to keep the tests for the stack impl. of `ReactTestRenderer` for now

gaearon commented on src/renderers/testing/fiber/ReactTestFiberComponent.js:
```diff
@@ -0,0 +1,45 @@
+
+var instanceCounter = 0;
+
+var ReactTestFiberComponent = {
```
Note: the separation between `ReactDOMFiber` and `ReactDOMFiberComponent` is just an artifact of how it was forked from `ReactDOMComponent` and that we still want to keep them in sync when possible. So feel free to abandon that convention (e.g. see `ReactNoop.js`)

spicyj commented:
> should I copy the ReactTestRenderer-test for ReactTestFiberRenderer?

It would be fine to branch on renderer in the test file like the ReactART tests do.

iamdustan commented:
Thanks for the feedback thus far.

Regarding refs I think I can reuse the `Renderer.findHostInstance` that already exists, though I don’t think I understand the value it is returning. For example: 


```jsx
    class Component extends React.Component {
      render() {
        return (
          <div className="purple">
            <div />
            <Child />
          </div>
        );
      }
    }

    class Child extends React.Component {
      render() {
        return (
          <div className="green" />
        );
      }
    }
    
    ReactTestRenderer.create(<Component />);
    const hostInstance = TestRenderer.findHostInstance(root);
    /* hostInstance ==>
{ type: 'div',
      children:
       [ { '$$typeof': Symbol(react.element),
           type: 'div',
           key: null,
           ref: null,
           props: {},
           _owner: [Object],
           _store: {} },
         { '$$typeof': Symbol(react.element),
           type: [Function: Child],
           key: null,
           ref: null,
           props: {},
           _owner: [Object],
           _store: {} },
         { type: 'div', children: null, props: {} },
         { type: 'div', children: null, props: [Object] } ],
      props: { className: 'purple' } }
*/
```

Is that expected? Why?

iamdustan commented:
Second issue is if I try to call `findHostInstance` on the (currently passed) Fiber to a ref callback I always get it as in an unmounted state.


```jsx
    class Component extends React.Component {
      render() {
        return (
          <div className="purple">
            <div />
            {/* this callback ref will throw */}
            <Child ref={fiber => console.log(ReactTestRenderer.findHostInstance(fiber)} />
          </div>
        );
      }
    }

    class Child extends React.Component {
      render() {
        return (
          <div className="green" />
        );
      }
    }
    ReactTestRenderer.create(<Component />);

// ref calback throws:
// Invariant Violation: Unable to find node on an unmounted component.
```

I would expect that to be mounted by the time the ref callback is called...

iamdustan commented:
Actually, ignore everything I’ve said. I think I see a path forward.

iamdustan commented:
Woohoo! Only 2 failing tests now!

I have ReactTestRenderer-test branching on the ReactDOMFeatureFlags useFiber switch now.

To support the `createNodeMock` function I needed to add a method to `ReactFiberHostContext` to support support hijacking the `attachRef` call in `ReactFiberCommitWork`. I’m a bit iffy about adding that since it seems intentional in the renderer API that that `instances` a renderer creates are exactly what the user gets in a ref.

With the two failing tests, I am curious if these are failing because the feature is incomplete or because behavior has changed subtly.

> warns correctly for refs on SFCs

Is this warning still to be implemented? If yes I would suspect it to “just work” here without any additional effort.

>  supports updates when using refs

The ref callback is not being called yet on unmount. This is another situation where I would expect this to just work. Is this not yet implemented or do I need to track down a bug in this renderer?

>  supports error boundaries

The order of ilfecycles is different here. Based on watching the error boundary work between @acdlite and @gaearon I suspect that this may be expected.

**Before**
```
        'Boundary render',
        'Angry render',
        'Boundary render',
        'Boundary componentDidMount',
```

**After**
```
        'Boundary render',
        'Angry render',
        'Boundary componentDidMount',
        'Boundary render',
```

iamdustan commented:
Just checked master. It looks like `warns correctly for refs on SFCs` is not implemented yet. That answers one of my questions.

gaearon commented:
@iamdustan Want to help with implementing that warning in master as a separate PR?

gaearon commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -86,7 +86,11 @@ module.exports = function<T, P, I, TI, C, CX>(
     const ref = finishedWork.ref;
     if (ref && (!current || current.ref !== ref)) {
-      ref(instance);
+      if (typeof config.getPublicInstance === 'function') {
```
Please destructure this from config earlier. We don't want to read that value every time. I think we could also make this required to keep contract strict.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,288 @@
+    parentHostContext : HostContext,
+    type : string,
+  ) : HostContext {
+    return {};
```
Please return the same object (emptyObject like in ReactNoop) so it doesn't push stack unnecessarily.

gaearon commented:
Can you please update this now that the SFC warning is in?

iamdustan commented:
The only failing test remaining is that it `supports updates when using refs`. I just copied that test to `refs-test` locally to see if the behavior already works in ReactDOMFiber with a slight change to account for `null` being passed on unmount (rather than the same node).

The test expects the ref to be called with `['div', 'div', 'span']` (the second div being on unmount), whilst the current behavior is to only be called with `['div', 'span']`.

My biggest concern is that being called twice wasn’t necessarily an intentional decision before, but the test was added to assert the behavior and not regress accidentally. I’m not certain how much effort should be put into maintaining this because I *think* it would require changing `ReactFiberCommitWork`s `detachRefIfNeeded` to call the `createNodeMock` method in some capacity.

@spicyj what thoughts or advise do you have since you wrote the original implementation?

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+    parentInstance.removeChild(child);
+  },
+
+  scheduleAnimationCallback: window.requestAnimationFrame,
```
I think this will throw in non-jsdom environment. Can you make it `setTimeout`? It seems like it wouldn't be used anyway, but at least we shouldn't throw.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+    rootContainerInstance : Container,
+  ) : boolean {
+    // console.log('finalizeInitialChildren');
+    // setInitialProperties(testElement, type, props, rootContainerInstance);
```
Please clean up the comments.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+        if (typeof child.toJSON === 'function') {
+          childrenJSON.push(child.toJSON());
+        } else if (typeof child.text !== 'undefined') {
+          childrenJSON.push();
```
Are pushing `undefined`? Why?

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  toJSON() {
+    // eslint-disable ignore the children
```
What are we ignoring specifically? Generally I'd like to avoid adding eslint ignores.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  update(type, props) {
+    // console.log('update', type, props, this.children);
```
Comments

gaearon commented on src/renderers/shared/fiber/ReactFiberReconciler.js:
```diff
@@ -47,6 +47,7 @@ export type HostConfig<T, P, I, TI, C, CX, CI> = {
 
   getRootHostContext(rootContainerInstance : C) : CX,
   getChildHostContext(parentHostContext : CX, type : T) : CX,
+  getPublicInstance(instance : I | TI) : any, // maybe add a PI (public instance type)?
```
Please do.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+    );
+  },
+
+  resetTextContent(testElement : Instance) : void {},
```
Nit: Could you please use a consistent style for empty methods? Either like this, or with `// Noop` comment as long as it's the same style everywhere.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+      text : text,
+      id: instanceCounter++,
+      rootContainerInstance,
+      toJSON: () => isNaN(+inst.text) ? inst.text : +inst.text,
```
I'm not sure this is right. For example even `"2"` would become `2`. Maybe this is unobservable implementation detail and we should just always use strings? Since that's what they get coerced to anyway.

iamdustan replied:
good point. this was to make one of the tests pass, but I think to be correct I need to capture the type when creating the TextInstance and only coerce back to a number if it came in as a number.

gaearon replied:
I don't think you'd actually get a number though in `createTextInstance`? As far as I can see it's already coerced by Fiber code by then. I think it actually makes more sense than exposing this to the renderer, and I'm fine with divergent behavior here (always treating text as strings).

gaearon commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -317,6 +343,7 @@ describe('ReactTestRenderer', () => {
     ]);
   });
 
+  // this is only passing in Fiber because refs aren't actually working
```
Is this comment outdated? It’s not telling me much: are refs broken in Fiber? Are they broken specifically for test renderer? What is our plan to fix them? Is this within the scope of this PR?

iamdustan replied:
outdated comment.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+  createNodeMock: Function;
+
+  constructor(rootID, createNodeMock) {
+    this.rootID = rootID;
```
I don't see `rootID` being used anywhere. Is it needed?

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+    }
+    var container = new TestContainer('<default>', createNodeMock);
+    var root = TestRenderer.createContainer(container);
+    if (root) {
```
When is it null?

iamdustan replied:
I don’t think ever, but flow was complaining that `TestRenderer.createContainer` may return `null`

gaearon replied:
You'd need to see why other renderers didn't need this. It probably inferred that `OpaqueNode` type can be null, maybe from some other method definition.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,332 @@
+  },
+
+  /* eslint-disable camelcase */
+  unstable_batchedUpdates() {
```
Can you just omit it? I don't think we want to expose it here.

iamdustan replied:
Oh yeah, forgot about this one. The stack-based ReactTestRenderer exposes this currently..

gaearon replied:
Exposing but throwing isn't much better. :-)
Let's either make it work or remove it.

iamdustan replied:
😇

...adding this to the PR checklist.

gaearon commented on src/renderers/shared/shared/__tests__/refs-test.js:
```diff
@@ -289,6 +289,28 @@ describe('ref swapping', () => {
+    }
+    ReactDOM.render(<Foo useDiv={true} />, container);
+    ReactDOM.render(<Foo useDiv={false} />, container);
+    expect(log).toEqual(['div', null, 'span']);
```
Ideally also unmount at the end and check we get another `null`?

iamdustan replied:
oh whoops, didn’t mean to actually commit this. This was testing the behavior difference for refs between the test-renderer and DOM renderer re https://github.com/facebook/react/pull/8628#issuecomment-271478813

gaearon commented:
I think the existing test for updates might be checking accidental implementation details.
It's not testing *refs themselves*, it's testing calls of `createNodeMock`.
It was added by @aweary in e43aaab2547870bc80c0b778604c5ee55b1d87f0, not by @spicyj.

In particular, I don't see why it calls `createNodeMock` for `div` twice. It might just be a bug in Stack version.

aweary commented:
Right now `createNodeMock` is always called when `getPublicInstance` is called, and when unmounting [`ReactOwner. removeComponentAsRefFrom`](https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactOwner.js#L113) calls it to see if the cached ref is equal to the current public instance. 

We can try caching the mock node on the instance and only calling `createNodeMock` when the instance is being mounted/updated.

iamdustan commented:
thanks @gaearon and @aweary. I added the caching and changed the test case accordingly in 087f1f66eb08475f259b65706423037f0efbf53e. All tests now pass.

Only remaining task is the `PI` public instance type parameter...

iamdustan commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -463,7 +463,7 @@ module.exports = function<T, P, I, TI, C, CX, CI>(
     const ref = finishedWork.ref;
     if (ref) {
-      const instance = getPublicInstance(finishedWork.stateNode);
+      const instance = (getPublicInstance(finishedWork.stateNode) : any);
```
the flow type for `instance` on master is `any`. Whilst adding the `PI` parameter does end up saying `instance` is `PI`, that fails flow due to ref expecting:

```
ref: null | (((handle : ?Object) => void) & { _stringRef: ?string }),
```

I suspect this could be improved, but for time constraints at the moment I simply dropped it back to equal the current inference.

gaearon commented on src/renderers/testing/stack/ReactTestRendererStack.js:
```diff
@@ -50,17 +50,21 @@ function getRenderedHostOrTextFromComponent(component) {
     this._renderedChildren = null;
     this._topLevelWrapper = null;
     this._hostContainerInfo = null;
+    this._nodeMock = UNSET;
```
Could we just assign it here instead? We already know the `element`.

iamdustan replied:
oh, would this need to be invalidated on `receiveComponent`?

gaearon replied:
Hmm, that's trickier. Probably not since nodes are supposed to be kept intact. I'm actually not sure the API is right now, maybe we should've just passed the `type` instead of the `element`. But what's done is done?

iamdustan replied:
welp, I’ve updated it to on `mountComponent` because that is the earliest point we can do so (that is when `hostContainerInfo` is passed in)

gaearon commented:
Ugh, GH collapsed my comment. Here it is: https://github.com/facebook/react/pull/8628#discussion_r95394305.

gaearon commented:
This is currently breaking `react-test-renderer` built package:

<img width="613" alt="screen shot 2017-01-10 at 18 30 33" src="https://cloud.githubusercontent.com/assets/810438/21819267/e53b614c-d762-11e6-8a54-6e81e15097a8.png">

(No, we don't have automated tests for them.)

This is because `ReactDOMFeatureFlags` is in `dom` folder so doesn't become part of the build.

iamdustan commented:
First two thoughts on how to fix this are:

* add `ReactDOMFeatureFlags` to the src array of `reactTestRenderer` in the gulpfile
* Create a `react-test-renderer/fiber.js` file for output and do some finagling to continue using `ReactDOMFeatureFlags.useFiber` for repo unit tests, but not break the build

gaearon commented:
I just pushed a commit that explicitly uses the Stack version (since that's what npm package should use).

iamdustan commented:
how do you feel about adding an entrypoint at `react-test-renderer/fiber` so consumers could opt-in to testing it early, too?

gaearon commented:
Added that too in that commit. (But don't expose it in `package.json` yet since ReactDOM doesn't expose it either. Not sure if intentional.)

iamdustan commented:
oh yeah. just saw it. :thumbsup:

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+    } else {
+      var childrenJSON = [];
+      this.children.forEach((child) => {
+        if (typeof child.toJSON === 'function') {
```
When would this ever be false? If I understand correctly `child` is also either an `Instance` or `TextInstance` which we have control over.

iamdustan replied:
disclaimer: this implementation was mostly an amalgamation of ReactNoop and the stack ReactTestRenderer so there is likely some unnecessary complexity from the copypasta.

gaearon replied:
Yea, just saying we should clean this up a little so the next person looking isn't more confused than we were.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+          childrenJSON.push(child.toJSON());
+        }
+      });
+      json.children = childrenJSON.length ? childrenJSON : null;
```
Wouldn't this whole block look simpler as

```js
      json.children = this.children.length ?
        this.children.map(child => child.toJSON()) :
        null;
```

?

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+    this.children = [];
+    this.rootContainerInstance = rootContainerInstance;
+
+    Object.defineProperty(this, 'id', { value: this.id, enumerable: false });
```
What is `id` field for? I tried removing it everywhere and tests still pass.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+  },
+
+  resetTextContent(testElement : Instance) : void {
+    // Noop
```
Why is this a no-op? It seems like there could be bugs related to this.
(e.g. DOM renderer implements this one.)

I think we should actually have `shouldSetTextContent` always return `false`.
AFAIK it only exists as an optimization to avoid creating additional Fibers for single-text-child divs.
But in case of test renderer it doesn't matter.

iamdustan replied:
what is the tdlr of this method? No tests currently require it so I just left it empty. Sounds like I’ll need to add a test case...

gaearon replied:
Try commenting its implementation out in DOM renderer and see what fails.
By grepping for method name, I found this:

```js
      if (nextEffect.effectTag & ContentReset) {
        config.resetTextContent(nextEffect.stateNode);
      }
```

By grepping for `ContentReset`, I found this:

```js
      // If we're switching from a direct text child to a normal child, or to
      // empty, we need to schedule the text content to be reset.
      workInProgress.effectTag |= ContentReset;
```

So it likely happens when we switch from a "direct text child" (with your implementation of `shouldSetTextContent`, it's the case with `<div>Text</div>` or `<div>{42}</div>`) to a "normal child" (e.g. `<div><p /></div>`). The purpose is to clear the text in this case.

I think you can avoid the whole problem by making `shouldSetTextContent` always return false.

gaearon replied:
Here is a failing test:

```js
  it('can update text nodes', () => {
    class Component extends React.Component {
      render() {
        return (
          <div>
            {this.props.children}
          </div>
        );
      }
    }

    var renderer = ReactTestRenderer.create(<Component>Hi</Component>);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      children: ['Hi'],
      props: {},
    });
    renderer.update(<Component>{['Hi', 'Bye']}</Component>);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      children: ['Hi', 'Bye'],
      props: {},
    });
    renderer.update(<Component>Bye</Component>);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      children: ['Bye'],
      props: {},
    });
    renderer.update(<Component>{42}</Component>);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      children: [42],
      props: {},
    });
    renderer.update(<Component><div /></Component>);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      children: [{
        type: 'div',
        children: null,
        props: {},
      }],
      props: {},
    });
```

Solved by returning `false` from `shouldSetTextContent`.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+    }
+    var container = new TestContainer(createNodeMock);
+    var root = TestRenderer.createContainer(container);
+    if (root) {
```
I don't see Flow errors removing this condition.

iamdustan replied:
great! sounds like it’s been magically resolved :)

iamdustan replied:
hmm...I can remove this one, but not those in the returned methods...

iamdustan replied:
oh duh. It’s because in `unmount`  we clean up our closured reference to `root`, but the consumer could still do:

```
const renderer = TestRenderer.create(<Comp />);
renderer.unmount();
renderer.update(<OtherComp />); // shouldn’t be possible because we unmounted.
```

I’ll need to look into what the current implementation does...

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+          return;
+        }
+        TestRenderer.updateContainer(null, root, null, () => {
+          container = null;
```
I think you shouldn't need this callback.
Since it uses synch scheduling, everything should happen synchronously anyway.
So you should be able to clean up immediately after calling `updateContainer`.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,325 @@
+  appendChild(child) {}
+  insertBefore(beforeChild, child) {}
+  removeChild(child) {}
+  toJSON() {}
```
I find these empty implementations odd.
Can you help me understand why they're unnecessary here?

What if I call top-level `create` with a fragment?

iamdustan replied:
The tldr is I originally wrote the `TestContainer` to hold the `createNodeMock` function and keep a pointer to it. This likely can be simplified. Top-level `create` with a fragment is likely broken right now.

I’ll look at unifying `TestComponent` and `TestContainer` tonight. Should I add a test case for top-level create?

gaearon replied:
Maybe. I'm pretty sure them being noops is wrong (since they do get called), but I'm not sure how to write a failing test. If you succeed with making a failing test against this then you'll know how to fix it.

iamdustan replied:
Hmm...this actually seems to *maybe* work okay already? Here is my current test case:

```jsx
    it('can update text nodes when rendered as root', () => {
      var Component = (props) => props.children;

      var renderer = ReactTestRenderer.create(<Component>Hi</Component>);
      expect(renderer.toJSON()).toEqual('Hi');
      renderer.update(<Component>{['Hi', 'Bye']}</Component>);
      expect(renderer.toJSON()).toEqual('Hi Bye');
      renderer.update(<Component>Bye</Component>);
      expect(renderer.toJSON()).toEqual('Bye');
      renderer.update(<Component>{42}</Component>);
      expect(renderer.toJSON()).toEqual(42);
      renderer.update(<Component><div /></Component>);
      expect(renderer.toJSON()).toEqual({
        type: 'div',
        children: null,
        props: {},
      });
    });
```

Though currently `<Component>{['Hi', 'Bye']}</Component>` `toJSON()` only returns `Hi` (instead of `Hi Bye`). I’m not actually certain what is expected at this point.

I’m going to add another test case later, family-time!

gaearon replied:
I'm confused. The test you're showing looks like my test about text nodes. But we were discussing root in this thread.

iamdustan replied:
Sorry, this is the root node test as I understand it: https://github.com/facebook/react/pull/8628/files#diff-417ca2789067ee4dcafab55a48aa03cbR532

```
    it('can render and update root fragments', () => {
      var Component = (props) => props.children;

      var renderer = ReactTestRenderer.create([
        <Component>Hi</Component>,
        <Component>Bye</Component>,
      ]);
      expect(renderer.toJSON()).toEqual('Hi');
      renderer.update(<div />);
      expect(renderer.toJSON()).toEqual({
        type: 'div',
        children: null,
        props: {},
      });
      renderer.update([<Component>{42}</Component>, <Component>The answer</Component>]);
      expect(renderer.toJSON()).toEqual(42);
    });
```

gaearon replied:
Yea, I don't really know how to trigger the failing case for containers. It's just odd to me that those methods are blank. I would expect that either they shouldn't be defined at all, or they should do something.

gaearon replied:
>expect(renderer.toJSON()).toEqual('Hi');

I don't think this looks right. We are rendering an array. Why do we get a single item as an output? I would expect it to return an array in Fiber.

iamdustan replied:
https://github.com/facebook/react/blob/e4733076498695c754740ff3a8f9d41c447b1a3e/src/renderers/dom/fiber/__tests__/ReactDOMFiber-test.js#L137-L158

The underlying implementation uses `TestRenderer.findHostInstance(root)`. From looking at what I believe are the corresponding DOM tests, only the *first* element in a fragment is returned.

iamdustan commented on src/renderers/testing/__tests__/ReactTestRenderer-test.js:
```diff
@@ -420,12 +445,106 @@ describe('ReactTestRenderer', () => {
+      var renderer = ReactTestRenderer.create(<Component>Hi</Component>);
+      expect(renderer.toJSON()).toEqual('Hi');
+      renderer.update(<Component>{['Hi', 'Bye']}</Component>);
+      expect(renderer.toJSON()).toEqual('Hi');
```
I’m not actually certain if this *should* be `Hi` or `Hi Bye`. I *think* that when given a fragment-like thing the first is the only one returned, but only spent 30 seconds on it so far.

gaearon replied:
It should be an array in Fiber. We should put the whole test behind a feature flag because it is Fiber specific. I think this demonstrates that root problem but maybe I'm wrong.

iamdustan commented:
Okay, I can get rid of all the `new` container instances for simple objects in exchange for this `toJSON` function. The reason this is still necessary is because:

* text instances need to be an object to be updated (`commitTextUpdate(instance, oldText, newText)` mutates the `instance`)
* need to be able to ignore the props in children `{props: {children: JSXProvidedChildrenWhichAreFibers}}`

This is the toJSON function:

```
const toJSON = child => {
  if (child.text) {
    return child.text;
  }

  /* eslint-disable no-unused-vars */
  /* ignoring the children prop and managing that ourselves*/
  const {children, ...props} = child.props;
  /* eslint-enable */
  const json = {
    type: child.type,
    props: props,
    children: null,
  };
  Object.defineProperty(json, '$$typeof', {value: Symbol.for('react.test.json')});
  if (child.children) {
    json.children = child.children.length
      ? child.children.map(toJSON)
      : null;
  }
  return json;
};
```

I can make that happen if you want. Flow is very unhappy with me at the moment, though so :shrug:

gaearon commented:
>I can make that happen if you want.

If it looks/works better, sure. If not, don’t bother. :-)
I’m not sure I fully understand the issues so it’s hard for me to say yes or no without seeing it.

iamdustan commented:
WIP commit d228c24 just pushed. I have three more flow issues to resolve to finish it off or I can revert it if it’s okay as is.

gaearon commented:
If you can figure out how to resolve them that would be great. Otherwise I can try to take a look.
I want to get this in today, in one way or another.

iamdustan commented:
:thumbsup: on it now.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -207,32 +142,44 @@ var TestRenderer = ReactFiberReconciler({
-    parentInstance.removeChild(child);
+    const index = parentInstance.children.indexOf(child);
+    if (index === -1) {
+      throw new Error('This child does not exist.');
```
Note sure those checks are important. They would indicate bugs in Fiber.
They exist in `ReactNoop` because it tests Fiber itself but doesn't look necessary here.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -246,8 +193,18 @@ var TestRenderer = ReactFiberReconciler({
   getPublicInstance(ref) {
-    const createNodeMock = ref.rootContainerInstance.createNodeMock;
-    return createNodeMock(ref);
+    if (typeof ref.text === 'string') {
```
Can we use a `$$typeof` switch here instead?

iamdustan replied:
how would that look? I could add `$$typeof` symbols to the container and text instances..

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -246,8 +193,18 @@ var TestRenderer = ReactFiberReconciler({
+    }
+
+    // this should not be possible
+    throw new Error('Attempted to getPublicInstance on an invalid ref.');
```
Hmm. Shouldn't `ref` be guaranteed to be either `Instance` or `Container`?

iamdustan replied:
I thought it was just `Instance` or `TextInstance`, though `Container` is also actually a possibility.

This was added for exhaustiveness because I couldn’t get flow to be okay without `else if` check just before this.

gaearon replied:
Okay. Switch on `$$typeof` with a throwing `default` would look nicer than duck-typing though.

iamdustan replied:
are there any examples of type checking with symbols?

gaearon replied:
Like [this](https://github.com/facebook/react/blob/1c43f9471773a5206279236c5e8b5b37d3e985cb/src/renderers/shared/fiber/ReactChildFiber.js#L397-L422)?

gaearon replied:
Not really important to use symbols there (if you don't leak those objects externally), you can just use an enum. Example [declaration](https://github.com/facebook/react/blob/1c43f9471773a5206279236c5e8b5b37d3e985cb/src/shared/ReactTypeOfWork.js#L15-L29) and [switch](https://github.com/facebook/react/blob/1c43f9471773a5206279236c5e8b5b37d3e985cb/src/renderers/shared/fiber/ReactFiberBeginWork.js#L596-L603).

iamdustan replied:
I added static strings before seeing this message. Okay?  	799e901

iamdustan commented:
I would like to take a brief moment to say thank you for your infinite patience and attention to detail in this PR @gaearon. :1st_place_medal:

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -28,7 +32,8 @@ type ReactTestRendererJSON = {|
   children: Array<Instance | TextInstance>,
-  createNodeMock: Function
+  createNodeMock: Function,
+  $$typeof: typeof CONTAINER_TYPE,
```
I think this could be just `$$typeof: CONTAINER_TYPE`?
Same below.

iamdustan replied:
without the `typeof` I get the following:

```
src/renderers/testing/ReactTestRendererFiber.js:36
 36:   $$typeof: CONTAINER_TYPE,
                 ^^^^^^^^^^^^^^ string. Ineligible value used in/as type annotation (did you forget 'typeof'?)
 36:   $$typeof: CONTAINER_TYPE,
                 ^^^^^^^^^^^^^^ CONTAINER_TYPE
```

gaearon commented:
This looks good to me. Thanks for sticking with it. 😄 
I’ll merge later today.

gaearon commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,293 @@
+  ) : void {
+    const index = parentInstance.children.indexOf(child);
+    if (index !== -1) {
+      this.children.splice(index, 1);
```
What is `this.children`?

iamdustan replied:
a bug. should be `parentInstance.children`.

gaearon commented:
Congratulations. 😄

sebmarkbage commented:
https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactFiberReconciler.js#L173 Should be updated to return `PI`.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -462,7 +463,7 @@ module.exports = function<T, P, I, TI, C, CX>(
     const ref = finishedWork.ref;
     if (ref) {
-      const instance = finishedWork.stateNode;
+      const instance = getPublicInstance(finishedWork.stateNode);
```
This is not type safe for the case `finishedWork.tag === ClassComponent`.

iamdustan replied:
how so? what can I do to rectify that?

sebmarkbage replied:
`finishedWork.stateNode` on a a class component will be the instance of the class. That doesn't match the generic types `I | TI`. I think it only accidentally works on runtime now because you assume they won't have a `.tag` field with the string `INSTANCE` so they fallback.

You can make separate branches by switching on `finishedWork.tag` and only use `getPublicInstance` on the branch on `HostComponent`.

sebmarkbage commented on src/renderers/testing/ReactTestRendererFiber.js:
```diff
@@ -0,0 +1,293 @@
+
+  useSyncScheduling: true,
+
+  getPublicInstance(inst) {
```
Hm. It's strange that you're allowed to avoid the type annotation here. It probably leads to some overly optimistic analysis by Flow. We should add a type annotation here. It's important to do in the renderers since otherwise the generic arguments can be inferred to be very weak.

iamdustan replied:
Related, how can I type the `createNodeMock` function so `PI` is properly inferred?
